### Import the "html_pitchfork_metadata.csv" which contains the artcile title, body text, posting date and author name of 3549 articles.

In [1]:
import pandas as pd
import nltk

pd.set_option('display.max_colwidth', 150)    # set the maximum column width to 150

df = pd.read_csv("html_pitchfork_metadata.csv", sep="\t")
df

article_title  \
0                Lapse in Passage   
1                       Origin EP   
2     Welcome to GStarr Vol. 1 EP   
3          A Muse in Her Feelings   
4        Norman Fucking Rockwell!   
...                           ...   
3544                   Blue World   
3545                  Laurel Hell   
3546                   Home Video   
3547                 King’s Mouth   
3548           Ants From Up There   

                                                                                                                                           article_bodytext  \
0      Mute Duo’s music is emblematic of their deep roots in Chicago.  Pedal steel guitarist Sam Wagster and drummer/multi-instrumentalist Skyler Rowe d...   
1      Kelly Moran’s Ultraviolet grew out of a period of writer’s block.  Her usual method of composition—painstakingly plotting every note on staff pap...   
2      Justin Rivera was 14 years old when he appeared on the second season of Lifetime’s 2016 music competition series The Rap Game as J. I.  the Princ...   
3      Daniel Daley has a voice so striking that it’s almost an affront he began his career as a rapper and songwriter; imagine having that gift and not...   
4      In 2017, Lana Del Rey stopped performing in front of the American flag.  Where the singer-songwriter born Elizabeth Grant had once stood onstage ...   
...                                                                                                                                                     ...   
3544   Sometimes a thing can be hidden without really hiding.  Blue World, a previously unissued cache of studio recordings by the classic John Coltrane...   
3545   Mitski sees the world with the ruthless, knowing gaze of someone who grew up too fast.  Her cutting acuity is part of what inspires rapture from ...   
3546   Lucy Dacus’ third album, Home Video, explores a slice of 2000s Christian youth culture from the perspective of a girl who lived through it.  It w...   
3547   If there’s been one constant in The Flaming Lips’ careening four-decade evolution from Okie goth-punk weirdos to festival-conquering freak show, ...   
3548   The Concorde jet: aeronautical disgrace, financial boondoggle, multinational embarrassment.  Hell of a metaphor, though.  Isaac Wood returns to i...   

            posting_date     article_author  
0         March 24, 2020  Jonathan Williger  
1           May 20, 2019   Philip Sherburne  
2          July 23, 2020    Alphonse Pierre  
3         April 22, 2020      Rawiya Kameir  
4      September 3, 2019         Jenn Pelly  
...                  ...                ...  
3544  September 30, 2019        Nate Chinen  
3545    February 2, 2022          Cat Zhang  
3546       June 24, 2021      Peyton Thomas  
3547       July 22, 2019      Stuart Berman  
3548    February 8, 2022          Ian Cohen  

[3549 rows x 4 columns]

### Tokenize the article body text in to sentences.

In [2]:
df["sentences"] = df.article_bodytext.apply(lambda x: nltk.sent_tokenize(x))
df["num_sentences"] = df.sentences.apply(lambda x: len(x))
df

article_title  \
0                Lapse in Passage   
1                       Origin EP   
2     Welcome to GStarr Vol. 1 EP   
3          A Muse in Her Feelings   
4        Norman Fucking Rockwell!   
...                           ...   
3544                   Blue World   
3545                  Laurel Hell   
3546                   Home Video   
3547                 King’s Mouth   
3548           Ants From Up There   

                                                                                                                                           article_bodytext  \
0      Mute Duo’s music is emblematic of their deep roots in Chicago.  Pedal steel guitarist Sam Wagster and drummer/multi-instrumentalist Skyler Rowe d...   
1      Kelly Moran’s Ultraviolet grew out of a period of writer’s block.  Her usual method of composition—painstakingly plotting every note on staff pap...   
2      Justin Rivera was 14 years old when he appeared on the second season of Lifetime’s 2016 music competition series The Rap Game as J. I.  the Princ...   
3      Daniel Daley has a voice so striking that it’s almost an affront he began his career as a rapper and songwriter; imagine having that gift and not...   
4      In 2017, Lana Del Rey stopped performing in front of the American flag.  Where the singer-songwriter born Elizabeth Grant had once stood onstage ...   
...                                                                                                                                                     ...   
3544   Sometimes a thing can be hidden without really hiding.  Blue World, a previously unissued cache of studio recordings by the classic John Coltrane...   
3545   Mitski sees the world with the ruthless, knowing gaze of someone who grew up too fast.  Her cutting acuity is part of what inspires rapture from ...   
3546   Lucy Dacus’ third album, Home Video, explores a slice of 2000s Christian youth culture from the perspective of a girl who lived through it.  It w...   
3547   If there’s been one constant in The Flaming Lips’ careening four-decade evolution from Okie goth-punk weirdos to festival-conquering freak show, ...   
3548   The Concorde jet: aeronautical disgrace, financial boondoggle, multinational embarrassment.  Hell of a metaphor, though.  Isaac Wood returns to i...   

            posting_date     article_author  \
0         March 24, 2020  Jonathan Williger   
1           May 20, 2019   Philip Sherburne   
2          July 23, 2020    Alphonse Pierre   
3         April 22, 2020      Rawiya Kameir   
4      September 3, 2019         Jenn Pelly   
...                  ...                ...   
3544  September 30, 2019        Nate Chinen   
3545    February 2, 2022          Cat Zhang   
3546       June 24, 2021      Peyton Thomas   
3547       July 22, 2019      Stuart Berman   
3548    February 8, 2022          Ian Cohen   

                                                                                                                                                  sentences  \
0     [ Mute Duo’s music is emblematic of their deep roots in Chicago., Pedal steel guitarist Sam Wagster and drummer/multi-instrumentalist Skyler Rowe ...   
1     [ Kelly Moran’s Ultraviolet grew out of a period of writer’s block., Her usual method of composition—painstakingly plotting every note on staff pa...   
2     [ Justin Rivera was 14 years old when he appeared on the second season of Lifetime’s 2016 music competition series The Rap Game as J. I.  the Prin...   
3     [ Daniel Daley has a voice so striking that it’s almost an affront he began his career as a rapper and songwriter; imagine having that gift and no...   
4     [ In 2017, Lana Del Rey stopped performing in front of the American flag., Where the singer-songwriter born Elizabeth Grant had once stood onstage...   
...                                                                                                                                                

### Merge all the sentences of each ariticle into a single list.

In [3]:
sentences = []
for sent in df.sentences:
    sentences = sentences + sent
print(len(sentences))

filtered_sentences = []
for s in sentences:
    if len(s) > 20:
        filtered_sentences.append(s)

print(len(filtered_sentences))

116275
114754


### Create a dataframe with the sentences and take a random sample of 2000 sentences.

In [4]:
df_sentences = pd.DataFrame(filtered_sentences, columns = ['sentences'])
df_sample_sentences = df_sentences.sample(n=2000, replace=False, random_state=0)
df_sample_sentences

sentences
30553   Even “The Blinding,” which is split between booming Swizz Beatz and AraabMuzik chant music and elegant Hit-Boy dinner party rap, finds its way ont...
2886                     It wasn’t entirely successful, though Kraviz’ search for new frontiers in live electronic music was, at the very least, provocative.
106070  It’s an old photograph of the artist’s mother in front of a body of water, her unique features erased, so that she becomes a shadow in the landsca...
31950   If they hadn’t so thoroughly flooded the market, perhaps there might have been more excitement around Wolf Parade’s 2017 reunion album Cry Cry Cry...
69192                    “Didascalies” underscores the idea that even the most solemn scenarios can devolve dramatically, a lesson implicit in The Favourite.
...                                                                                                                                                       ...
87001   Coming after recent behind-the-scenes collaborations with artists like Rex Orange County, PJ Morton, and The Free Nationals, Music prominently fea...
60478   Rather than recrimination and blame, Rose offers a more measured appreciation of their time together, recalling the specifics of the relationship ...
39786                                           It’s a bit peppier than Slowdive and Orbital’s early-’90s crystallizations of noise, dub, and Detroit techno.
89226                                                                                                             Sign up for the 10 to Hear newsletter here.
59639                       On heavily processed tracks like “Hex” and “Viva Shona,” ping-ponging electronic dots and dashes nearly crowd him out of the mix.

[2000 rows x 1 columns]

### Export the Random sample sentences into a csv file.

In [5]:
df_sample_sentences.sentences.to_csv("pitchfork_randomsample_sentences.csv", index=False)

### pitchfork_randomsample_sentences.csv is downloaded and each sentence in this file is labelled with one of the following sentiments.

- 1 - Positive
- 0 is Neutral
- -1 is Negative

## Load Labelled Sentences csv file

In [6]:
df_labelled = pd.read_csv("pitchfork_randomsample_sentences_Labelled.csv", sep="\t")
df_labelled

sentences  \
0     Even “The Blinding,” which is split between booming Swizz Beatz and AraabMuzik chant music and elegant Hit-Boy dinner party rap, finds its way ont...   
1                      It wasn’t entirely successful, though Kraviz’ search for new frontiers in live electronic music was, at the very least, provocative.   
2     It’s an old photograph of the artist’s mother in front of a body of water, her unique features erased, so that she becomes a shadow in the landsca...   
3     If they hadn’t so thoroughly flooded the market, perhaps there might have been more excitement around Wolf Parade’s 2017 reunion album Cry Cry Cry...   
4                      “Didascalies” underscores the idea that even the most solemn scenarios can devolve dramatically, a lesson implicit in The Favourite.   
...                                                                                                                                                     ...   
1995  Coming after recent behind-the-scenes collaborations with artists like Rex Orange County, PJ Morton, and The Free Nationals, Music prominently fea...   
1996  Rather than recrimination and blame, Rose offers a more measured appreciation of their time together, recalling the specifics of the relationship ...   
1997                                          It’s a bit peppier than Slowdive and Orbital’s early-’90s crystallizations of noise, dub, and Detroit techno.   
1998                                                                                                            Sign up for the 10 to Hear newsletter here.   
1999                      On heavily processed tracks like “Hex” and “Viva Shona,” ping-ponging electronic dots and dashes nearly crowd him out of the mix.   

      sentiment  
0             1  
1             0  
2            -1  
3            -1  
4            -1  
...         ...  
1995          0  
1996         -1  
1997          0  
1998          0  
1999         -1  

[2000 rows x 2 columns]

In [7]:
df_labelled= df_labelled.drop_duplicates(keep= "first") #To remove the duplicate sentences
df_labelled

sentences  \
0     Even “The Blinding,” which is split between booming Swizz Beatz and AraabMuzik chant music and elegant Hit-Boy dinner party rap, finds its way ont...   
1                      It wasn’t entirely successful, though Kraviz’ search for new frontiers in live electronic music was, at the very least, provocative.   
2     It’s an old photograph of the artist’s mother in front of a body of water, her unique features erased, so that she becomes a shadow in the landsca...   
3     If they hadn’t so thoroughly flooded the market, perhaps there might have been more excitement around Wolf Parade’s 2017 reunion album Cry Cry Cry...   
4                      “Didascalies” underscores the idea that even the most solemn scenarios can devolve dramatically, a lesson implicit in The Favourite.   
...                                                                                                                                                     ...   
1994                                                                               “First Time,” one of the strongest tracks, is both spacious and patient.   
1995  Coming after recent behind-the-scenes collaborations with artists like Rex Orange County, PJ Morton, and The Free Nationals, Music prominently fea...   
1996  Rather than recrimination and blame, Rose offers a more measured appreciation of their time together, recalling the specifics of the relationship ...   
1997                                          It’s a bit peppier than Slowdive and Orbital’s early-’90s crystallizations of noise, dub, and Detroit techno.   
1999                      On heavily processed tracks like “Hex” and “Viva Shona,” ping-ponging electronic dots and dashes nearly crowd him out of the mix.   

      sentiment  
0             1  
1             0  
2            -1  
3            -1  
4            -1  
...         ...  
1994          1  
1995          0  
1996         -1  
1997          0  
1999         -1  

[1944 rows x 2 columns]

### Get the value counts sentiment variable.

In [8]:
df_labelled.sentiment.value_counts()

 0    678
 1    654
-1    612
Name: sentiment, dtype: int64

### Modeling the Data

In [9]:
# We import the Tfidvectorizer to create a document term matrix that gives us our X(feature variables) and y(target variable)
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(use_idf=True, norm="l2", stop_words="english", max_df=0.7)

x = vectorizer.fit_transform(df_labelled.sentences) # Our X column with shape = 1944, 10121

y= df_labelled.sentiment # Our y column with shape = 1944

In [10]:
# To randomly split the dataset into 2 sets for 80% training and 20% testing. we import from sklearn
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split (x, y, test_size = 0.2, random_state = 0)

In [11]:
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((1555, 10121), (1555,), (389, 10121), (389,))

### Building the Models

#### 1. K-Nearest Neighbors

In [12]:
# We import the kNeighbors classifer

from sklearn.neighbors import KNeighborsClassifier

from sklearn.model_selection import cross_val_score # We split the data repeatedly & train multiple models from these new training sets
knc = KNeighborsClassifier(n_neighbors=1) #we set our parameter to 1
knc.fit(x_train, y_train) # We fit the model to the 80% training data
# To get the higest possible variable max score from for random parameters.
score_max = 0              # Score_max is a temoporay variable to store the max score 

for p in [1, 5, 10, 30]: #we set our parameters using these random numbers
    model = KNeighborsClassifier(n_neighbors=p)
    scores = cross_val_score(model, x_train, y_train, cv=5)
    print(f"k = {p}: {scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}\n")
    
    if scores.mean() > score_max:
        score_max = scores.mean()
        p_best = p         # Param_best is a temoporay variable to store the best parameter 
        
print(f"Highest score : {round(score_max, 3)} when k = {p_best}") 

k = 1: [0.35369775 0.35369775 0.36655949 0.36334405 0.34405145]
0.356, 0.008

k = 5: [0.34083601 0.35691318 0.35369775 0.31189711 0.35691318]
0.344, 0.017

k = 10: [0.34405145 0.33762058 0.37942122 0.37299035 0.37299035]
0.361, 0.017

k = 30: [0.37299035 0.36655949 0.34083601 0.40514469 0.33118971]
0.363, 0.026

Highest score : 0.363 when k = 30


In [13]:
from sklearn.metrics import confusion_matrix, classification_report
y_true = y_test
y_pred = knc.predict(x_test)
print(classification_report(y_true, y_pred)) 

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00       132
           0       0.30      1.00      0.46       115
           1       0.00      0.00      0.00       142

    accuracy                           0.30       389
   macro avg       0.10      0.33      0.15       389
weighted avg       0.09      0.30      0.13       389



/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


##### To build the final model with our optimal parameter

In [14]:
def train_test(X_train, X_test, y_train, y_test, cls):
    cls.fit(X_train, y_train)
    
    y_true = y_test
    y_pred = cls.predict(X_test)
    
    print(f"Train accuracy score: {round(cls.score(X_train, y_train), 3)}")
    print(f"Test accuracy score: {round(cls.score(X_test, y_test), 3)}\n")
    print(confusion_matrix(y_true, y_pred))
    print()
    print(classification_report(y_true, y_pred, zero_division=0))
    
    return cls

In [15]:
print(f"k = {p_best}")
knc = KNeighborsClassifier(n_neighbors=p_best)#initialize a new model with the previous parameter
%time knc = train_test(x_train, x_test, y_train, y_test, knc)

k = 30
Train accuracy score: 0.45
Test accuracy score: 0.329

[[36 69 27]
 [28 63 24]
 [32 81 29]]

              precision    recall  f1-score   support

          -1       0.38      0.27      0.32       132
           0       0.30      0.55      0.38       115
           1       0.36      0.20      0.26       142

    accuracy                           0.33       389
   macro avg       0.34      0.34      0.32       389
weighted avg       0.35      0.33      0.32       389

CPU times: user 109 ms, sys: 11.6 ms, total: 120 ms
Wall time: 118 ms


###### Making Predictions

In [16]:
text1 = "Dancing all night to the fave band has me feeling so good this morning. It was a good time!"
text2 = "I will not be bullied into hiding my face because I dislike His songs."
text3 = "Today, I became a fan of Classic music. Move over jazz!"
text4 = "I am so sad"
text5 = "They are playing football nicely."
text6 = "The officer arrested the culprits."
text7 = "We were not sad when he moved away."
text8 = "He will not have received enough score by the summer."
text9 = "I had not been cooking any meal when she came to my home."
text10 = "Mary and Samantha took the bus"

new_texts = [text1, text2, text3, text4, text5, text6, text7, text8, text9, text10]
x_new = vectorizer.transform(new_texts)

knc.predict(x_new) #assign x_new as the new varaible to predict for K-NN

array([ 0, -1, -1, -1,  1,  0,  0,  1,  0,  0])

In [17]:
print(f"k = {p_best}")
knc = KNeighborsClassifier(n_neighbors=p_best)#initialize a new model with the previous parameter
%time knc = train_test(x_train, x_test, y_train, y_test, knc)

k = 30
Train accuracy score: 0.45
Test accuracy score: 0.329

[[36 69 27]
 [28 63 24]
 [32 81 29]]

              precision    recall  f1-score   support

          -1       0.38      0.27      0.32       132
           0       0.30      0.55      0.38       115
           1       0.36      0.20      0.26       142

    accuracy                           0.33       389
   macro avg       0.34      0.34      0.32       389
weighted avg       0.35      0.33      0.32       389

CPU times: user 105 ms, sys: 8.28 ms, total: 113 ms
Wall time: 113 ms


In [18]:
summary = {}
summary["k-NNs"] = round(knc.score(x_test, y_test), 3) # Add K-NNs to the Summary dictionary

In [19]:
knc.predict(x_new)

array([ 0, -1, -1, -1,  1,  0,  0,  1,  0,  0])

#### 2. Logistics Regression

In [20]:
from sklearn.linear_model import LogisticRegression # We import the Logistics Regression (LR) classifer

lr = LogisticRegression(random_state=0) # We set our random state to 0 to control the degree of randomnessscores = cross_val_score(lr, x_train, y_train, cv=5)
print(f"{scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}")

[0.37299035 0.36655949 0.34083601 0.40514469 0.33118971]
0.363, 0.026


In [21]:
%time lr = train_test(x_train, x_test, y_train, y_test, lr)

Train accuracy score: 0.992
Test accuracy score: 0.365

[[33 60 39]
 [18 64 33]
 [27 70 45]]

              precision    recall  f1-score   support

          -1       0.42      0.25      0.31       132
           0       0.33      0.56      0.41       115
           1       0.38      0.32      0.35       142

    accuracy                           0.37       389
   macro avg       0.38      0.37      0.36       389
weighted avg       0.38      0.37      0.36       389

CPU times: user 4.87 s, sys: 17.6 s, total: 22.5 s
Wall time: 312 ms


In [22]:
summary["Logistic Regression"] = round(lr.score(x_test, y_test), 3) # Add the Logistics Regression to the Summary dictionary

In [24]:
lr.predict(x_new)

array([ 0, -1,  0])

#### 3. Multinomial Naive Bayes

In [25]:
from sklearn.naive_bayes import MultinomialNB    # We import the Multinomial Bayes classifer

mnb = MultinomialNB()
scores = cross_val_score(mnb, x_train, y_train, cv=5)
print(f"{scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}")

[0.35691318 0.37942122 0.35369775 0.37942122 0.34405145]
0.363, 0.014


In [26]:
%time mnb = train_test(x_train, x_test, y_train, y_test, mnb) #assign it to mb

Train accuracy score: 0.987
Test accuracy score: 0.37

[[31 65 36]
 [16 68 31]
 [20 77 45]]

              precision    recall  f1-score   support

          -1       0.46      0.23      0.31       132
           0       0.32      0.59      0.42       115
           1       0.40      0.32      0.35       142

    accuracy                           0.37       389
   macro avg       0.40      0.38      0.36       389
weighted avg       0.40      0.37      0.36       389

CPU times: user 4.8 ms, sys: 2.06 ms, total: 6.87 ms
Wall time: 6.36 ms


In [27]:
summary["Multinomial Naive Bayes"] = round(mnb.score(x_test, y_test), 3) # Add MNC to the Summary dictionary

In [28]:
mnb.predict(x_new)

array([0, 1, 1])

#### 4. Decision Trees

In [29]:
from sklearn.tree import DecisionTreeClassifier    # We import the Decision Tree classifer

dtc = DecisionTreeClassifier(random_state=0)

scores = cross_val_score(dtc, x_train, y_train, cv=5)
print(f"{scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}")

[0.37942122 0.31189711 0.36012862 0.35691318 0.36977492]
0.356, 0.023


In [30]:
%time dtc = train_test(x_train, x_test, y_train, y_test, dtc)

Train accuracy score: 0.998
Test accuracy score: 0.37

[[42 56 34]
 [26 59 30]
 [40 59 43]]

              precision    recall  f1-score   support

          -1       0.39      0.32      0.35       132
           0       0.34      0.51      0.41       115
           1       0.40      0.30      0.35       142

    accuracy                           0.37       389
   macro avg       0.38      0.38      0.37       389
weighted avg       0.38      0.37      0.37       389

CPU times: user 402 ms, sys: 0 ns, total: 402 ms
Wall time: 402 ms


In [31]:
summary["Decision Trees"] = round(dtc.score(x_test, y_test), 3) # Add DT to the Summary dictionary

In [32]:
dtc.predict(x_new)

array([0, 0, 1])

#### 5. Random Forest

In [33]:
from sklearn.ensemble import RandomForestClassifier #Import the classifier

rfc = RandomForestClassifier(random_state=0)

scores = cross_val_score(rfc, x_train, y_train, cv=5)
print(f"{scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}")

[0.35691318 0.32154341 0.34083601 0.38263666 0.36012862]
0.352, 0.02


In [34]:
%time rfc = train_test(x_train, x_test, y_train, y_test, rfc)

Train accuracy score: 0.998
Test accuracy score: 0.35

[[ 19  96  17]
 [  8  98   9]
 [ 17 106  19]]

              precision    recall  f1-score   support

          -1       0.43      0.14      0.22       132
           0       0.33      0.85      0.47       115
           1       0.42      0.13      0.20       142

    accuracy                           0.35       389
   macro avg       0.39      0.38      0.30       389
weighted avg       0.40      0.35      0.29       389

CPU times: user 2.4 s, sys: 1.54 ms, total: 2.4 s
Wall time: 2.4 s


In [35]:
summary["Random Forest"] = round(rfc.score(x_test, y_test), 3) # Add RF to the Summary dictionary


In [36]:
rfc.predict(x_new)

array([1, 0, 0])

#### 6. Linear Support Vector Machines (SVM)

In [37]:
from sklearn.svm import LinearSVC

lsvc = LinearSVC(random_state=0)

score_max = 0

for param in [0.01, 0.03, 0.1, 0.3, 1, 3, 10]: #try different values
    model = LinearSVC(C=param, random_state=0)
    scores = cross_val_score(model, x_train, y_train, cv=5)
    print(f"C = {param}: {scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}\n")
    
    if scores.mean() > score_max: #to get the max average score and its corresponding value
        score_max = scores.mean()
        param_best = param
        
print(f"Highest score : {round(score_max, 3)} when C = {param_best}")

C = 0.01: [0.36012862 0.36012862 0.36334405 0.36334405 0.36334405]
0.362, 0.002

C = 0.03: [0.33118971 0.35691318 0.36012862 0.36655949 0.36334405]
0.356, 0.013

C = 0.1: [0.35369775 0.36334405 0.36655949 0.40836013 0.33762058]
0.366, 0.023

C = 0.3: [0.36655949 0.36334405 0.35048232 0.37620579 0.33118971]
0.358, 0.016

C = 1: [0.36655949 0.35369775 0.35048232 0.35691318 0.32154341]
0.35, 0.015

C = 3: [0.35369775 0.36334405 0.36334405 0.35369775 0.32475884]
0.352, 0.014

C = 10: [0.33440514 0.36655949 0.36012862 0.35048232 0.32154341]
0.347, 0.017

Highest score : 0.366 when C = 0.1


In [38]:
print(f"C = {param_best}")  #re-initialize our model, pass the parameter to a new name, then run
lsvc = LinearSVC(C=param_best)
%time lsvc = train_test(x_train, x_test, y_train, y_test, lsvc)

C = 0.1
Train accuracy score: 0.985
Test accuracy score: 0.37

[[28 68 36]
 [16 72 27]
 [21 77 44]]

              precision    recall  f1-score   support

          -1       0.43      0.21      0.28       132
           0       0.33      0.63      0.43       115
           1       0.41      0.31      0.35       142

    accuracy                           0.37       389
   macro avg       0.39      0.38      0.36       389
weighted avg       0.39      0.37      0.35       389

CPU times: user 9.72 ms, sys: 0 ns, total: 9.72 ms
Wall time: 8.74 ms


In [39]:
summary["Linear SVMs"] = round(lsvc.score(x_test, y_test), 3) # Add L-SVM to the Summary dictionary

In [40]:
lsvc.predict(x_new)

array([0, 0, 0])

#### 7. Kernelized Support Vector Machines (KSVMs)

In [41]:
from sklearn.svm import SVC

svc = SVC(C=1, kernel="rbf", gamma="scale", random_state=0)
score_max = 0

for parameter in [0.01, 0.03, 0.1, 0.3, 1, 3, 10]: #Using different parameters to get the best score
    model = SVC(C=param, kernel="rbf", gamma="scale", random_state=0)
    scores = cross_val_score(model, x_train, y_train, cv=5)
    print(f"C = {parameter}: {scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}\n")
    
    if scores.mean() > score_max:
        score_max = scores.mean()
        parameter_best = parameter
        
print(f"Highest score : {round(score_max, 3)} when C = {parameter_best}")

C = 0.01: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 0.03: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 0.1: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 0.3: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 1: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 3: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

C = 10: [0.36012862 0.35048232 0.36977492 0.38585209 0.31189711]
0.356, 0.025

Highest score : 0.356 when C = 0.01


In [42]:
print(f"C = {parameter_best}")
svc = SVC(C=parameter_best)
%time svc = train_test(x_train, x_test, y_train, y_test, svc)

C = 0.01
Train accuracy score: 0.362
Test accuracy score: 0.296

[[  0 132   0]
 [  0 115   0]
 [  0 142   0]]

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00       132
           0       0.30      1.00      0.46       115
           1       0.00      0.00      0.00       142

    accuracy                           0.30       389
   macro avg       0.10      0.33      0.15       389
weighted avg       0.09      0.30      0.13       389

CPU times: user 953 ms, sys: 0 ns, total: 953 ms
Wall time: 952 ms


In [43]:
summary["Kernelized SVMs"] = round(svc.score(x_test, y_test), 3) # Add K-SVMs to the Summary dictionary

In [44]:
svc.predict(x_new)

array([0, 0, 0])

#### 8. Neural Networks

In [45]:
from sklearn.neural_network import MLPClassifier

mlpc = MLPClassifier(hidden_layer_sizes=(10, ), random_state=0)

score_max = 0

for p_meter in [10, 30, 100]:
    model = MLPClassifier(hidden_layer_sizes=(p_meter, ), random_state=0)
    scores = cross_val_score(model, x_train, y_train, cv=5)
    print(f"hidden_layer_size = {p_meter}: {scores}\n{round(scores.mean(), 3)}, {round(scores.std(), 3)}\n")
    
    if scores.mean() > score_max:
        score_max = scores.mean()
        p_meter_best = p_meter
        
print(f"Highest score : {round(score_max, 3)} when hidden_layer_sizes = {p_meter_best}")

/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/rpandurangan/.local/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/rpandurangan/.local/li

hidden_layer_size = 10: [0.34083601 0.34083601 0.35048232 0.36334405 0.33118971]
0.345, 0.011

hidden_layer_size = 30: [0.34405145 0.37299035 0.36334405 0.35691318 0.31511254]
0.35, 0.02

hidden_layer_size = 100: [0.34405145 0.34726688 0.34405145 0.37299035 0.33762058]
0.349, 0.012

Highest score : 0.35 when hidden_layer_sizes = 30


In [46]:
print(f"hidden_layer_size = {p_meter_best}")
mlpc = MLPClassifier(hidden_layer_sizes=(p_meter_best, ), random_state=0)
%time mlpc = train_test(x_train, x_test, y_train, y_test, mlpc)

hidden_layer_size = 30
Train accuracy score: 0.998
Test accuracy score: 0.388

[[44 51 37]
 [22 55 38]
 [29 61 52]]

              precision    recall  f1-score   support

          -1       0.46      0.33      0.39       132
           0       0.33      0.48      0.39       115
           1       0.41      0.37      0.39       142

    accuracy                           0.39       389
   macro avg       0.40      0.39      0.39       389
weighted avg       0.40      0.39      0.39       389

CPU times: user 54.7 s, sys: 3min 27s, total: 4min 22s
Wall time: 4.1 s


In [47]:
summary["Neural Networks"] = round(mlpc.score(x_test, y_test), 3) # Add NN to the Summary dictionary

In [48]:
mlpc.predict(x_new)

array([0, 1, 1])

### Selecting the Best Model

In [49]:
summary

{'k-NNs': 0.329,
 'Logistic Regression': 0.365,
 'Multinomial Naive Bayes': 0.37,
 'Decision Trees': 0.37,
 'Random Forest': 0.35,
 'Linear SVMs': 0.37,
 'Kernelized SVMs': 0.296,
 'Neural Networks': 0.388}

Our overall performance using these 8 classification models was low. The best of the models is the Neural Network which gives a score of 0.388. This is low and may indicate:

- Methodologies
    - The set of optimal parameters yieling the best performance may have not been found. 
    - The classification algorithm yiendling the best performance may have not been found. 
    - The features used (sentences), may not serve as predictors, i.e., have no predictive power. It is important to consider adjusting the features (although there's no guarantee that it would work if none of the features have predictive power.)
- Data
    - The quality of the labeled data is not very good, i.e., the labels are not very accurate. 
    - The number of labeled samples is not sufficient.